File: javascript/beer-song/beer-song.js
function BeerSong(){

  this.verse = function(number){
    number = (number === 0) ? 'No more' : number;
    let pluralOne = number === 1 ? '' : 's';
    let pluralTwo = number > 2 ? 's' : '';
    pluralTwo = number === 1 ? 's' : pluralTwo;
    let secondNum = number > 1 ? (number-1) : "no more";
    let itOne = number > 1 ? "one" : "it";
    let firstBit = `${number} bottle${pluralOne} of beer on the wall, ${typeof number === 'number' ? number : number.toLowerCase()} bottle${pluralOne} of beer.\n`

    let secondBit = `Take ${itOne} down and pass it around, ${secondNum} bottle${pluralTwo} of beer on the wall.\n`;

    secondBit = number === 'No more' ? 'Go to the store and buy some more, 99 bottles of beer on the wall.\n' : secondBit;

    return (firstBit + secondBit);
  }

  this.sing = function(verse1, verse2=0){
    let array = [];
    while (verse1 >= verse2){
      array.push(this.verse(verse1));
      verse1--;
    }
    return array.join('\n');
  }
}

module.exports = BeerSong;

File: ruby/beer-song/beer_song.rb
module BookKeeping
  VERSION = 2
end

class BeerSong

  def verse(number)
    number = (number == 0) ? 'No more' : number
    second_num = number.to_i > 1 ? (number-1) : "no more"
    it_one = number.to_i > 1 ? "one" : "it"
    first_bit = "#{number} bottle#{number == 1 ? '' : 's'} of beer on the wall, #{number.class == 'Fixnum' ? number : number.to_s.downcase} bottle#{number == 1 ? '' : 's'} of beer.\n"
    second_bit = "Take #{it_one} down and pass it around, #{second_num} bottle#{second_num != 1 ? 's' : ''} of beer on the wall.\n"
    second_bit = number == "No more" ? "Go to the store and buy some more, 99 bottles of beer on the wall.\n" : second_bit
    first_bit + second_bit
  end

  def verses(num1, num2)
    num1.downto(num2).map { |num| verse(num) }.join("\n")
  end

  def lyrics
    verses(99, 0)
  end 

end
